Add index disk usage API (#74051)
elastic/elasticsearch: docs/reference/indices/diskusage.asciidoc

the parameter `run_expensive_tasks` must be set to `true` in order for the task to be
performed.
Defaults to `false`.

`flush`::
(Optional, Boolean)
If `true`, the target index will be flushed before the analyzing process starts.
If `false`, then uncommitted data might not be included in the result.
Defaults to `true`.

[[analyze-index-disk-usage-api-example]]
==== {api-examples-title}

The following request analyzes the disk usage of index `my-index-000001`.

[source,console]
--------------------------------------------------
POST /my-index-000001/_index_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:my_index]

The API returns the following response:

[source,console-response]
--------------------------------------------------
{
    "_shards": {
        "total": 1,
        "successful": 1,
        "failed": 0
    },
    "my-index-000001": {
        "_all": {
            "total": "925.4mb",
            "total_in_bytes": 970370607,
            "terms": "38.6mb",
            "terms_in_bytes": 40541738,
            "postings": "65.7mb",
            "postings_in_bytes": 68963453,

>             "proximity": "11.6mb",

Reviewer: It is challenging because some bytes record information about two different things. E.g. some vints for doc IDs steal a bit to know whether the freq is greater than 1, and some vints for positions steal a bit to know whether there is a payload associated with this position.

>             "proximity": "11.6mb",

Author: +1, I will add this option in a follow-up.

[[indices-disk-usage]]
=== Analyze index disk usage API
++++
<titleabbrev>Analyze index disk usage</titleabbrev>
++++

experimental[]

Analyzes the disk usage of each field of an index. This API

> might not support indices created in previous {es} versions.

Reviewer: Also maybe there should be a note that this API only works well on large indices, since it ignores the disk footprint of metadata (field info, and more generally all the things that Lucene reads upon opening the index).

[source,console]
--------------------------------------------------
POST /my-index-000001/_index_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:my_index]

[[analyze-index-disk-usage-api-request]]
==== {api-request-title}

`POST /<index>/_index_disk_usage`

[[analyze-index-disk-usage-api-request-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have the `manage`
<<privileges-list-indices,index privilege>> for the target index or index alias.

[[analyze-index-disk-usage-api-path-params]]
==== {api-path-parms-title}

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index]
+
By default, you must explicitly name the indices you are analyzing the disk usage.
To specify indices to close with `_all`, `*`, or other wildcard
expressions, change the `action.destructive_requires_name` setting to `false`.
You can update this setting in the `elasticsearch.yml` file

> or using the <<cluster-update-settings,cluster update settings>> API.

Author: I added this because I prefer this API to accept a single API. Should we still enforce this, but not a destructive action? If so then we can drop `_index` from `_index_disk_usage` API?

[[analyze-index-disk-usage-api-query-params]]
==== {api-query-parms-title}

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
+
Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
+
Defaults to `open`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards]

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]

`run_expensive_tasks`::
(Optional, Boolean)
Since analyzing per field disk usage of indices is expensive and resource-intensive,
the parameter `run_expensive_tasks` must be set to `true` in order for the task to be
performed.
Defaults to `false`.

`flush`::
(Optional, Boolean)
If `true`, the target index will be flushed before the analyzing process starts.
If `false`, then uncommitted data might not be included in the result.
Defaults to `true`.

[[analyze-index-disk-usage-api-example]]
==== {api-examples-title}

The following request analyzes the disk usage of index `my-index-000001`.

[source,console]
--------------------------------------------------

> POST /my-index-000001/_index_disk_usage?run_expensive_tasks=true

Author: I preferred `_disk_usage` over `_index_disk_usage` but if the index is not specified then `POST /_disk_usage` can be understood as analyzing the disk usage of the cluster.

> POST /my-index-000001/_index_disk_usage?run_expensive_tasks=true

Author: Ok, I changed `_index_disk_usage` to `_disk_usage`

--------------------------------------------------
// TEST[setup:my_index]

The API returns the following response:

[source,console-response]
--------------------------------------------------
{
    "_shards": {
        "total": 1,
        "successful": 1,
        "failed": 0
    },
    "my-index-000001": {
        "_all": {
            "total": "925.4mb",
            "total_in_bytes": 970370607,
            "terms": "38.6mb",
            "terms_in_bytes": 40541738,
            "postings": "65.7mb",
            "postings_in_bytes": 68963453,
            "proximity": "11.6mb",
            "proximity_in_bytes": 12253834,
            "stored_fields": "623.5mb",
            "stored_fields_in_bytes": 653843556,
            "doc_values": "125.7mb",
            "doc_values_in_bytes": 131882253,
            "points": "59.9mb",
            "points_in_bytes": 62885773,
            "norms": "0b",
            "norms_in_bytes": 0,
            "vectors": "0b",

>             "vectors_in_bytes": 0

Author: +1, I pushed https://github.com/elastic/elasticsearch/pull/74051/commits/4d86ecba2c55c077f9109aa00527f6f58625c72a to rename "vectors" to "term vectors".

<<privileges-list-indices,index privilege>> for the target index, data stream,
or alias.

[[analyze-index-disk-usage-api-path-params]]
==== {api-path-parms-title}

`<target>`::
(Required, string) Comma-separated list of data streams, indices, and aliases
used to limit the request. It's recommended to execute this API with a single
index (or the latest backing index of a data stream) as the API consumes
resources significantly.

[[analyze-index-disk-usage-api-query-params]]
==== {api-query-parms-title}

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
+
Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
+
Defaults to `open`.

`flush`::
(Optional, Boolean) If `true`, the API performs a flush before analysis. If
`false`, the response may not include uncommitted data. Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]

`run_expensive_tasks`::
(Required, Boolean) Analyzing field disk usage is resource-intensive. To use the

> API, this parameter must `true`. Defaults to `false`.

Reviewer 2: +1. Thanks for spotting this @ywelsch.

> API, this parameter must `true`. Defaults to `false`.

Author: fixed in https://github.com/elastic/elasticsearch/pull/74051/commits/d453b76455e8ebcf7d2dadc9efb66b152b38f2ef

--------------------------------------------------
POST /my-index-000001/_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:messages]

The API returns:

[source,console-response]
--------------------------------------------------
{
    "_shards": {
        "total": 1,
        "successful": 1,
        "failed": 0
    },
    "my-index-000001": {
        "store_size": "929mb", <1>
        "store_size_in_bytes": 974192723,
        "all_fields": {
            "total": "928.9mb", <2>
            "total_in_bytes": 973977084,
            "inverted_index": "107.8mb",
            "inverted_index_in_bytes": 113128526,
            "stored_fields": "623.5mb",
            "stored_fields_in_bytes": 653819143,
            "doc_values": "125.7mb",
            "doc_values_in_bytes": 131885142,
            "points": "59.9mb",
            "points_in_bytes": 62885773,
            "norms": "2.3kb",
            "norms_in_bytes": 2356,
            "term_vectors": "2.2kb",
            "term_vectors_in_bytes": 2310
        },
        "fields": {
            "_id": {
                "total": "49.3mb",
                "total_in_bytes": 51709993,

>                 "inverted_index": "29.7mb",

Author: adjusted in https://github.com/elastic/elasticsearch/pull/74051/commits/cac588a08a73a45cef7c0102c3f56c971a47c0fd

[[indices-disk-usage]]
=== Analyze index disk usage API
++++
<titleabbrev>Analyze index disk usage</titleabbrev>
++++

experimental[]

Analyzes the disk usage of each field of an index or data stream.
This API might not support indices created in previous {es} versions.
The result of a small index can be inaccurate as some parts of an index
might not be analyzed by the API.

[source,console]
--------------------------------------------------
POST /my-index-000001/_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:messages]

[[analyze-index-disk-usage-api-request]]
==== {api-request-title}

`POST /<target>/_disk_usage`

[[analyze-index-disk-usage-api-request-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have the `manage`
<<privileges-list-indices,index privilege>> for the target index, data stream,
or alias.

[[analyze-index-disk-usage-api-path-params]]
==== {api-path-parms-title}

`<target>`::
(Required, string) Comma-separated list of data streams, indices, and aliases
used to limit the request. It's recommended to execute this API with a single

> index (or the latest backing index of a data streams) as the API consumes

Reviewer 2: ```suggestion
index (or the latest backing index of a data stream) as the API consumes
```

resources significantly.

[[analyze-index-disk-usage-api-query-params]]
==== {api-query-parms-title}

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
+
Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
+
Defaults to `open`.

`flush`::
(Optional, Boolean) If `true`, the API performs a flush before analysis. If
`false`, the response may not include uncommitted data. Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]

`run_expensive_tasks`::

> (Required, Boolean) Analyzing field disk usage is resource-intensive. To use the

Reviewer 3: Just to clarify, does it mean that `_disk_usage` will not run at all if `run_expensive_tasks` is false? If yes, than what's the need of this parameter?  Is `run_expensive_tasks` parameter is some kind extra defence mechanism?

Author: >  Is run_expensive_tasks parameter is some kind extra defence mechanism

Yes, that's the purpose. It requires an acknowledge from users.

--------------------------------------------------
{
    "_shards": {
        "total": 1,
        "successful": 1,
        "failed": 0
    },
    "my-index-000001": {
        "index_size": "929mb", <1>
        "index_size_in_bytes": 974192723,
        "all_fields": {
            "total": "928.9mb", <2>
            "total_in_bytes": 973977084,
            "inverted_index": "107.8mb",
            "inverted_index_in_bytes": 113128526,
            "stored_fields": "623.5mb",
            "stored_fields_in_bytes": 653819143,
            "doc_values": "125.7mb",
            "doc_values_in_bytes": 131885142,
            "points": "59.9mb",
            "points_in_bytes": 62885773,
            "norms": "2.3kb",
            "norms_in_bytes": 2356,
            "term_vectors": "2.2kb",
            "term_vectors_in_bytes": 2310
        },
        "fields": {
            "_id": {
                "total": "49.3mb",
                "total_in_bytes": 51709993,
                "inverted_index": "29.7mb",
                "inverted_index_in_bytes": 31172745,
                "stored_fields": "19.5mb", <3>
                "stored_fields_in_bytes": 20537248,
                "doc_values": "0b",

>                 "doc_values_in_bytes": 0,

Reviewer 3: Do you think it would be useful to have an option to skip items with 0 bytes?

Author: I had this filter before but removed. I don't have a strong opinion on this. I am happy with either way.

`false`, the response may not include uncommitted data. Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]

`run_expensive_tasks`::
(Required, Boolean) Analyzing field disk usage is resource-intensive. To use the
API, this parameter must be set to `true`. Defaults to `false`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards]

[[analyze-index-disk-usage-api-example]]
==== {api-examples-title}

[source,console]
--------------------------------------------------
POST /my-index-000001/_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:messages]

The API returns:

[source,console-response]
--------------------------------------------------
{
    "_shards": {
        "total": 1,
        "successful": 1,
        "failed": 0
    },
    "my-index-000001": {
        "store_size": "929mb", <1>
        "store_size_in_bytes": 974192723,
        "all_fields": {
            "total": "928.9mb", <2>
            "total_in_bytes": 973977084,

>             "inverted_index": {

Reviewer: should we return an object for all data structures for consistency?

Reviewer 4: At least in the field usage API I'm inclined not to, as many of these would just have a single value below them. Here there would be two (X and X_in_bytes), so it might just make this overly hierarchical.

I have to admit that I'm not 100% happy with the `inverted_index` section either, but expect our thinking to evolve as we add more details to this API here in follow-ups.

[[indices-disk-usage]]
=== Analyze index disk usage API
++++
<titleabbrev>Analyze index disk usage</titleabbrev>
++++

experimental[]

Analyzes the disk usage of each field of an index. This API
might not support indices created in previous {es} versions.
The result of a small index can be inaccurate as some parts
of an index might not be analyzed by the API.

> 

Author: ++, I pushed https://github.com/elastic/elasticsearch/pull/74051/commits/fd81b30ff8fe065ee48f58ac435af7a61e4f91b7.

[[indices-disk-usage]]
=== Analyze index disk usage API
++++
<titleabbrev>Analyze index disk usage</titleabbrev>
++++

experimental[]

> Analyzes the disk usage of each field of an index. This API

Reviewer 2: ```suggestion
Analyzes the disk usage of each field of an index or data stream. This API
```

might not support indices created in previous {es} versions.
The result of a small index can be inaccurate as some parts
of an index might not be analyzed by the API. It's recommended
to execute this API with a single index (or the latest backing
index of a data streams) as the API consumes resources significantly.

[source,console]
--------------------------------------------------
POST /my-index-000001/_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:messages]

[[analyze-index-disk-usage-api-request]]
==== {api-request-title}

> `POST /<index>/_disk_usage`

Reviewer 2: Since the API can target data streams. I'd use `<target>` instead of `<index>`.

```suggestion
`POST /<target>/_disk_usage`
```

Reviewer 5: `index` is used many places in the API to mean "index, index alias, or data stream" (`search`, `index`, `create`, `indices.get_mapping`, etc) and `target` is typically used when taking something from one source and applying it elsewhere (see `indices.shrink`, `indices.clone`, `indices.split`) so unsure if `target` would be consistent with other areas of our API spec?

Definitely agree with the change to document that this API can take data streams as a parameter.

Reviewer 2: > `index` is used many places in the API to mean "index, index alias, or data stream" (`search`, `index`, `create`, `indices.get_mapping`, etc)

We use `<target>` in the docs for APIs that can target both data streams and indices:

* [Search API](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-search.html)
* [Index API](https://www.elastic.co/guide/en/elasticsearch/reference/master/docs-index_.html)
* [Get mapping](https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-get-mapping.html)

Using `<index>` as a parameter that can accept data streams is confusing as there are a few APIs that only accept an index, such as the [create index API](https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-create-index.html).

Reviewer 5: Ah my apologies! I read this as a comment on the API spec. Disregard me and carry on!

[[analyze-index-disk-usage-api-request-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have the `manage`

> <<privileges-list-indices,index privilege>> for the target index or index alias.

Reviewer 2: ```suggestion
<<privileges-list-indices,index privilege>> for the target index, data stream,
or alias.
```

[[analyze-index-disk-usage-api-path-params]]
==== {api-path-parms-title}

> include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index]

Reviewer 2: As noted above, I'd change this to `<target>`.
```suggestion
`<target>`::
(Required, string) Comma-separated list of data streams, indices, and aliases
used to limit the request. Supports wildcards (`*`).
```

[[analyze-index-disk-usage-api-query-params]]
==== {api-query-parms-title}

> include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
> +
> Defaults to `true`.
> 
> include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
> +
> Defaults to `open`.
> 
> include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]
> 
> include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards]
> 
> include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]
> 
> `run_expensive_tasks`::
> (Optional, Boolean)
> Since analyzing the per field disk usage of indices is expensive and resource-intensive,
> the parameter `run_expensive_tasks` must be set to `true` in order for the task to be
> performed.
> Defaults to `false`.
> 
> `flush`::
> (Optional, Boolean)
> If `true`, the target index will be flushed before the analyzing process starts.
> If `false`, then uncommitted data might not be included in the result.
> Defaults to `true`.

Reviewer 2: I sorted these in alphabetical order. I also made some minor edits to the `flush` and `run_expensive_tasks` params.
```suggestion
include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=allow-no-indices]
+
Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=expand-wildcards]
+
Defaults to `open`.

`flush`::
(Optional, Boolean) If `true`, the API performs a flush before analysis. If
`false`, the response may not include uncommitted data. Defaults to `true`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailable]

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=timeoutparms]

`run_expensive_tasks`::
(Required, Boolean) Analyzing field disk usage is resource-intensive. To use the
API, this parameter must `true`. Defaults to `false`.

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards]
```

[[analyze-index-disk-usage-api-example]]
==== {api-examples-title}

> The following request analyzes the disk usage of index `my-index-000001`.
> 

Reviewer 2: I don't think this is needed.
```suggestion
```

[source,console]
--------------------------------------------------
POST /my-index-000001/_disk_usage?run_expensive_tasks=true
--------------------------------------------------
// TEST[setup:messages]

> The API returns the following response:

Reviewer 2: Nit:
```suggestion
The API returns:
```

[source,console-response]
--------------------------------------------------
{
    "_shards": {
        "total": 1,
        "successful": 1,
        "failed": 0
    },
    "my-index-000001": {

>         "index_size": "929mb", <1>

Reviewer 4: In other places we call this `store.size`, so I'm wondering if this should be `store_size`?

Author: Adjusted in https://github.com/elastic/elasticsearch/pull/74051/commits/3af606cce371d63413ab5c74cebade99e494a25c.

        "index_size_in_bytes": 974192723,
        "all_fields": {
            "total": "928.9mb", <2>
            "total_in_bytes": 973977084,

>             "terms": "107.8mb",

Reviewer 4: I reached out to Adrien to discuss this:
- In his proposal `inverted_index` would be a container for all the fields that belong to the inverted index structures, namely terms, postings, frequencies, positions, offsets, payloads. The field usage API would then show the information as:

```
"inverted_index" : {
    "terms" : 1,
    "frequencies" : 1,
    "positions" : 0,
    "offsets" : 0,
    "payloads" : 0
},
"doc_values" : 1,
"points" : 0,
...
```

while the disk usage API would group the information under the same  "inverted_index"  section, but with possibly differently named subfields (or even no subfield at all in a first step). In a first step, the disk usage API could just show a total here, not a more refined view (as that would require introducing more concepts/names for what's currently the combined terms and frequencies disk usage). This would already help users make a choice between "index: true | false" for a field. 
- Another option we discussed would be for field usage to carry the  "terms", "frequencies", "positions" etc. fields, and the disk usage would then use combined field names such as "terms+frequencies" to make it clear that the metric shown is displaying disk usage from both of these combined. The field names would look as follows:

```
field_usage:

- "terms"
- "frequencies"
- "positions"
- "offsets"
- "payloads"
- "stored_fields"
- "doc_values"
- "points"
- "norms"
- "term_vectors"

disk_usage:

- "terms+frequencies"
- "positions+offsets+payloads"
- "stored_fields"
- "doc_values"
- "points"
- "norms"
- "term_vectors"
```

- It's also possible to combine both proposals by putting these combined field names under an "inverted_index" section.

Finally we discussed that disk usage could have multiple modes of executing, a quicker and a more expensive one, where the more expensive one could provide a more detailed breakdown of frequencies vs terms storage for example by rewriting the index.

>             "terms": "107.8mb",

Author: @ywelsch @jpountz Thanks for discussing it. I like the name `inverted_index` most and in this iteration, it just shows the total (i.e., no subfields).

> Finally we discussed that disk usage could have multiple modes of executing, a quicker and a more expensive one, where the more expensive one could provide a more detailed breakdown of frequencies vs terms storage for example by rewriting the index.

+1. I will work on it in a follow-up.

                "stored_fields_in_bytes": 0,
                "doc_values": "0b",
                "doc_values_in_bytes": 0,
                "points": "0b",
                "points_in_bytes": 0,
                "norms": "2.3kb",
                "norms_in_bytes": 2356,
                "term_vectors": "2.2kb",
                "term_vectors_in_bytes": 2310
            },
            "context.keyword": {...},
            "message": {...},
            "message.keyword": {...}
        }
    }
}
--------------------------------------------------
// TESTRESPONSE[s/: \{\.\.\.\}/: $body.$_path/]
// TESTRESPONSE[s/: (\-)?[0-9]+/: $body.$_path/]
// TESTRESPONSE[s/: "[^"]*"/: $body.$_path/]

<1> The total size of only analyzed shards of the index. This size can be
smaller than the store size of the index because only one copy of each shard
is examined. Also, some shards can have more than one index commit due to
snapshots or recoveries.

<2> The total size of fields of the analyzed shards of the index. This total
is usually smaller than the index size specified in <1> as some small metadata
files are ignored and some parts of data files might not be scanned by the API.

> <3> The stored size of the field `_id`

Reviewer 2: Nit:
```suggestion
<3> The stored size of the `_id` field.
```

> <4> The stored size of the field `_source`. As stored fields are stored

Reviewer 2: ```suggestion
<4> The stored size of the `_source` field. As stored fields are stored
```

together in a compressed format, the estimated sizes of stored fields are

> best efforts and can be inaccurate. The stored size of the field `_id`
> is likely underestimated while the field `_source` is overestimated.

Reviewer 2: ```suggestion
best efforts and can be inaccurate. The stored size of the `_id` field
is likely underestimated while the `_source` field is overestimated.
```
